specs/CustomDownload.md

```diff
@@ -0,0 +1,528 @@
+  ///
+  /// If the event is not handled, downloads will complete normally with the
+  /// default dialog shown.
+  ///
+  /// \snippet ScenarioCustomDownloadExperience.cpp CustomDownloadExperience
+  HRESULT add_DownloadStarting(
+    [in] ICoreWebView2StagingDownloadStartingEventHandler* eventHandler,
+    [out] EventRegistrationToken* token);
+
+  /// Remove an event handler previously added with `add_DownloadStarting`.
+  HRESULT remove_DownloadStarting(
+      [in] EventRegistrationToken token);
+}
+
+[uuid(53b676ec-c3b1-4804-996a-c72c16b09efb), object, pointer_default(unique)]
+interface ICoreWebView2DownloadStartingEventHandler : IUnknown
+{
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke(
+      [in] ICoreWebView2_3* sender,
+      [in] ICoreWebView2DownloadStartingEventArgs* args);
+}
+
+/// Event args for the `DownloadStarting` event.
+[uuid(c7e95e2f-f789-44e6-b372-3141469240e4), object, pointer_default(unique)]
+interface ICoreWebView2DownloadStartingEventArgs : IUnknown
+{
+  /// Returns the `ICoreWebView2DownloadItem` for the download that has started.
+  HRESULT DownloadItem([out, retval] ICoreWebView2DownloadItem** downloadItem);
+
+  /// The host may set this flag to cancel the download. If canceled, the
+  /// download save dialog will not be displayed regardless of the
+  /// ShouldDisplayDefaultDownloadDialog property.
+  [propget] HRESULT Cancel([out, retval] BOOL* cancel);
+
+  /// Sets the `Cancel` property.
+  [propput] HRESULT Cancel([in] BOOL cancel);
+
+  /// The path to the file. If setting the path, the host should ensure that it
+  /// is an absolute path, including the file name. If the directory does not
+  /// exist, it will be created.
+  [propget] HRESULT ResultFilePath([out, retval] LPWSTR* resultFilePath);
+
+  /// Sets the `ResultFilePath` property.
+  [propput] HRESULT ResultFilePath([in] LPCWSTR resultFilePath);
+
+  /// Returns an `ICoreWebView2Deferral` object.  Use this operation to
+  /// complete the event at a later time.
+  HRESULT GetDeferral([out, retval] ICoreWebView2Deferral** deferral);
+}
+
+/// Represents a download process. Gives access to the download's metadata
+/// and supports a user canceling, pausing, or resuming the download.
+[uuid(0C4A07B1-B610-459F-A574-EC253D5EC40D), object, pointer_default(unique)]
+interface ICoreWebView2DownloadItem : IUnknown
+{
+  /// The unique ID of the download.
+  [propget] HRESULT Id([out, retval] UINT32* id);
+
+  /// The URI of the download.
+  [propget] HRESULT Uri([out, retval] LPWSTR* uri);
+
+  /// The Content-Disposition header value from HTTP response.
+  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);
+
+  /// MIME type of the downloaded content.
+  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);
+
+  /// The size of the download in total number of expected bytes.
+  [propget] HRESULT DownloadSizeInBytes([out, retval] UINT64* downloadSizeInBytes);
```

Comment:
It is the download size before file compression. Should I add Expected?

Reply:
What do you mean by file compression? Do you mean the encodings applied to the HTTP body as its is transferred? I'm not worried about that. I think this has to be the expected size based on the initial HTTP response's headers and that it could be missing or incorrect compared to what we actually end up getting? If so then add Expected. Does this value change as the download goes on or completes?

Reply:
Yes you're right, it comes from content-length header. Added Expected